lib/ProductOpener/ContributionKnowledgePanels.pm

```diff
@@ -0,0 +1,158 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::ContributionKnowledgePanels - Generate knowledge panels around contribution
+
+=head1 SYNOPSIS
+
+This is a subpart of KnowledgPanels where we concentrate around contribution informations
+
+=cut
+
+package ProductOpener::ContributionKnowledgePanels;
+
+use ProductOpener::PerlStandards;
+use Exporter qw< import >;
+
+use Log::Any qw($log);
+
+BEGIN {
+	use vars qw(@ISA @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT_OK = qw(
+		&create_contribution_card_panel
+		&create_data_quality_panel
+	);    # symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK;
+
+use ProductOpener::KnowledgePanels qw(create_panel_from_json_template);
+use ProductOpener::Tags qw(:all);
+
+use Encode;
+use Data::DeepAccess qw(deep_get);
+
+=head2 create_contribution_card_panel ( $product_ref, $target_lc, $target_cc, $options_ref )
+
+Creates a knowledge panel card that contains all knowledge panels related to contribution
+
+=head3 Arguments
+
+=head4 product reference $product_ref
+
+Loaded from the MongoDB database, Storable files, or the OFF API.
+
+=head4 language code $target_lc
+
+Returned attributes contain both data and strings intended to be displayed to users.
+This parameter sets the desired language for the user facing strings.
+
+=head4 country code $target_cc
+
+We may display country specific recommendations from health authorities, or country specific scores.
+
+=head4 options reference $options_ref
+
+=cut
+
+sub create_contribution_card_panel ($product_ref, $target_lc, $target_cc, $options_ref) {
+
+	$log->debug("create contribution card panel", {code => $product_ref->{code}}) if $log->is_debug();
+
+	my @panels = ();
+	for my $tag_type (qw(data_quality_errors data_quality_warnings data_quality_info)) {
+		# we need to create it first because it can condition contribution panel display
+		my $created = create_data_quality_panel($tag_type, $product_ref, $target_lc, $target_cc, $options_ref);
+		push(@panels, $created) if $created;
+	}
+	if (@panels) {
+		my $panel_data_ref = {quality_panels => \@panels,};
+		create_panel_from_json_template("contribution_card",
+			"api/knowledge-panels/contribution/contribution_card.tt.json",
+			$panel_data_ref, $product_ref, $target_lc, $target_cc, $options_ref);
+	}
+	return;
+}
+
+# private sub to add tag names to data_quality_tags_by_actions: we just iterate to add tag name
```

Comment:
I think it's easier and cleaner to call the function in the template where we need it

e.g. in other templates we have things like [% display_taxonomy_tag("ingredients", ingredient) %]

The function is available by default for all templates.

Reply:
Right !